### ISSUE TITLE
feat: improve datetime usage with inputs on desktop

### ISSUE THREAD

**Issue by andreas-aeschlimann** [ionic-team/ionic-framework#24477]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already include this feature request, without success.

### Describe the Feature Request

The new date and time pickers are graphically beautiful, but the user experience is horrible, especially on the desktop. The complaints from our PWA users were that strong, we had to implement a new version of a date and time picker. As Ionic is heading towards a solution for mobile AND desktop, it should support the desktop scenario in a better way.

These are the issues with the components:
- A full datepicker looks clumsy in a form because it is so huge. This is a problem on mobile and desktop.
- We should have a modal version out of the box (which solves the problem above). In the docs, there is an example with a modal, but it needs additional CSS tweaking to make it work.
- When in a modal, there should be buttons to close it. In the given example, a user has to click outside of the modal to close it which is very unintuitive
- All scrollable options (months and time picker) are really bad on the Desktop. It is basically impossible to understand how it works for first-time users
- The time picker has no confirm button. Again, the user has to click outside of the overlay to save the time

----

Ideas to solve the issues:

1) In my opinion, a date picker should be an ion-input by default and open with a click in a modal as given below. Dates are relatively easy to insert with a keyboard. Hence, all you need is a good validator and a nice picker for mobile in addition.

This is my implemention using an ion-item with a ion-label, ion-input and ion-modal. Unfortunately, it was a lot of work and I think Ionic should offer something like this out of the box.

<img width="602" alt="Screenshot 2021-12-22 at 21 04 39" src="https://user-images.githubusercontent.com/2174826/147149079-2660fd6d-9ae7-46e9-bc0b-c8013c45534c.png">

Once you click the icon on the right, a modal opens. Note the buttons below to confirm and close the modal. This is really important for the user experience, otherwise you would need to click outside the modal or press escape, which does not make sense.

<img width="337" alt="Screenshot 2021-12-22 at 21 11 19" src="https://user-images.githubusercontent.com/2174826/147149844-14879091-5e6d-414a-ab71-ce45059b8f45.png">

Because the time picker is that awful on desktop, we had to split it from the date component. For the time picker one could use ion-input combined with ion-picker, which works quite well on both desktop and mobile. The validation for the input is much simpler with times of course, because you may only have values 0–23 and 0–59.

2) Another beautiful approach has been seen in the known calendar app "WeekCal". Once the item with the datetime is clicked, a picker opens. Once you click somewhere else, the picker hides again (similar to an accordion).

This only works for running calendars though, because you would not want to select a date far away from now. The year is missing and there would probably not be enough space on mobile to show the "D-M-Y H:m" format.

<img src="https://user-images.githubusercontent.com/2174826/147147838-6af99007-1365-4322-b0b9-f2b34e13c0c6.jpeg" width="350px" />

----

If you are absolutely fond of the current datetime component, we could implement a new component (for example: ion-input-datetime) that implements scenario 1.

That said, is there currently a way to use the current ion-picker in the way as I described in number 2? I think inline template use of the picker is not possible yet.

### Describe the Use Case

Desktop users could interact with the `ion-datetime` component in an acceptable way.

I will be happy to contribute some design ideas with some code, if requested.

### Additional Information

We love Ionic and I hope that my criticism was not too harsh. I just wanted to help you understand the impression of real-world users of these components.

**Comment by sean-perkins** [comment]
Hello @andreas-aeschlimann thanks for the feedback. I do think it's important to consider the hard work that many individuals on the team have put into the design and implementation of the date picker when wording feedback. Labels such as "awful" can be perceived as an insult to those individuals' work and I know your intention was not to be insulting 👍 

When looking at UX patterns for date time, I know that the team evaluated both platform specific patterns (iOS/Android) as well as what the web defaults are.

If you look at the `datetime-locale` implementation for the web for example: https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_input_type_datetime-local you will notice there are no confirmation or cancel buttons in the selection. 

Also looking at the human interface guidelines for iOS: https://developer.apple.com/design/human-interface-guidelines/ios/controls/pickers/ you'll see that in the compact appearance, the calendar does not have these buttons as well. Only in the wheels display is their actions and those are attributed to the higher modal and not the time picker itself. 

I strongly believe we should look at the Ionic UI components as the building blocks to your application's interface. We should offer good default components, with appropriate extensions available, to accomplish the vastly different implementation patterns that exist. 

If we had provided documented examples of a few common implementation patterns with date time that you could then encapsulate in your own component in your framework of choosing, would that have felt like a good interaction to you?

**Comment by andreas-aeschlimann** [comment]
Hello @sean-perkins, thanks for your feedback. Again, this was just an observation and nothing personal. :)

While I know that the datetime picker follows the current mobile guidelines, they do not work for users without a touch screen. Also, considering the fact that Apple has changed its calendar datetime picker with every of the last few OS versions, I believe their feedback has also been mixed. :D 

So what can we do? We should either provide a third whole new style option (ios, md and desktop), but that would raise a bigger feature request than this one. Or we could add some additional config options to the current components (like showing scroll bars for desktop users, buttons to click/close). 

When looking at your w3schools example, I see that you are right about the buttons. But isn't this exactly the reason every CSS framework (like Bootstrap) has its own datetime implementation? The default behavior of the browser is just awful. There is one very important difference from your w3schools example to the Ionic implementation: It has an input for the keyboard users. This means, even though the modal is not beautifully designed, I can interact with the field in an efficient way.

As a developer, I would like to use `<ion-datetime>` in a form in the same way as `<ion-input>`, but I cannot unless I accept to have no input possibility.

So about your last sentences: We could help users by publishing examples from my scenario 1 in the docs indeed. To achieve such a solution, one needs to write a lot of code. Especially connecting an ion-input with an ion-datetime is tricky due to validation, synchronisation. And if you want to reuse the code, you need to encapsulate it into a new component (note: I am on Angular) that implements a ValueAccessor, etc.

There are other problems with custom implementations. For example, even the Ionic example in the source example relies on a specific pixel values:

https://github.com/ionic-team/ionic-docs/blob/4b3edc8d8d2daf752b2de033bcd27a2d38244e04/static/demos/api/datetime/index.html#L33-L37

This means that any simple pixel change in Ionic's underlying datetime component will break the GUI of all users that implement this solution. This is why I would prefer a solution where we could create a datetime-modal out of the box.

**Comment by sean-perkins** [comment]
@andreas-aeschlimann I think you raise very valid and real concerns regarding the UX of the date picker that will require a larger team communication after the New Year.

I do see the value in lower-effort quicker wins, like options to introduce common UX patterns like rendering confirmation buttons and/or enabling scroll indications. 

I could also see value in encapsulating the new `ion-datetime` within `ion-input`, similar to how you can do: `<ion-input type="datetime">` but with our custom web component instead of falling back to the browser. 

I had not considered the extra complexity of value accessors in Angular, I do agree that's a less ideal path if implementers have to manage that. 

The value of the `ion-datetime` in its current state, is that it's a freshly architected component in v6, with less overhead complexity by delegating a lot of the timezone conditions to other libraries that manage that well. This allows us additional capacity to closer evaluate UX patterns and shape the component in future minor releases of v6. 

Appreciate the feedback 👍

**Comment by andreas-aeschlimann** [comment]
I like the idea of a new `ion-input` type. It could be "datetime" and "date" and "time".

Then, the component could be containing the real input and the button to open a modal with an ion-datetime within.

I am happy to assist you with further discussions. I could also post some snippets of my approach.

**Comment by sean-perkins** [comment]
Hello everyone following this issue :wave: 

Spoke with the team and have a few different thoughts to share:

1. The UX pattern of introducing confirm/cancel buttons to the date & time picker does not appear to be a common pattern noticeable across other popular UI libraries. We believe a better pattern would be similar to the "New Event" date/time selection in iOS.

<video src="https://user-images.githubusercontent.com/13732623/148251978-98464fb6-f9f0-4c47-8cee-87aed6f16a64.MP4"></video>

There is an open PR for introducing a compact style to the time picker that we believe nicely overlaps with a missing requirement to accomplishing that UI. We will share more around this piece after we have a moment to catch up and fully review that PR (#24403)

If you are focused more within a desktop appearance, we still are recommending splitting the controls, but having the date and time appearance independent with their own confirm/cancel buttons:

```html
<ion-datetime presentation="time" show-default-buttons="true">
  ...
</ion-datetime>
```

To summarize we don't currently think _nested_ confirm/cancel actions within the date & time display makes sense for this component, but there are tangible alternatives to achieve a confirm/cancel UX with independent display of date and independent display of time.

2. We believe the introduction of the v6 datetime component with an `ion-input` makes a lot of sense, especially as it's a common pattern to offer a text input with a calendar display. I'd be happy to either repurpose this feature request into that or create a new one.

3. Scroll indications will still require more review/discovery. Unable to provide direction on that at this time.


Let me know if you have any feedback or concern around some of these thoughts above ☝️

**Comment by andreas-aeschlimann** [comment]
@sean-perkins this sounds great, thank you for your post.

1. The "accordion" version is interesting. It looks graphically pleasing and is also quite efficient. I thought about implementing this by myself, but I could not figure out how to use the ion-picker within an ion-item. The current behavior is overlay-only I think. If Ionic can provide a solution as mentioned, it would be great.

2. The text input is essential for Desktop users and I strongly recommend it. It maybe could still be used in combination of the behavior described in the first point.

3. Scroll indicators are not really necessary if you have the first two points. It would be mostly useful for Desktop users. But then again, they would most likely use the text input in most cases.

**Comment by JesseB0rn** [comment]
Just to leave some of my thoughts and feedback I got from my users here:

- The time picker really is not that accessible to less tech savy folks (My dad is the product manager of my startup, he still  could barely set a time.) I think there should be an option to use iOS 14's interface (A keypad to type the time in)
- Some of my users also voiced harsher formulations against the UX

I will open another pull request later with the iOS 14 UX/UI, as I think that would also cover the desktop/PWA use-case.

**Comment by 1beb** [comment]
I also have anecdotal evidence from older (60+) users of apps that I've built. The time picker specifically (spinner) was challenging for them to control when selecting a specific time. We ended up going with text entry, especially when collecting dates that were known to the user (like birthdays)

**Comment by liamdebeasi** [comment]
Hi everyone,

I wanted to give an update on the primary pain points noted here and the Ionic team's plan for addressing each:


1. **When opening the time picker in the popover, it is not clear how to confirm the time because there are no buttons.** 

As noted in https://github.com/ionic-team/ionic-framework/issues/24477#issuecomment-1005881804, we recommend splitting the date and time controls out for this use case. This follows a pattern commonly seen on iOS where the date and time pickers are split into separate form controls. Doing this allows developers to provide clear instructions on how to confirm date and time values.

2. **Adding a button to open a datetime from an input requires a lot of code overhead to display the datetime in a modal.**

We are going to explore the idea of making this easier to manage via https://github.com/ionic-team/ionic-framework/issues/24402. The idea here is that you could pass in `ion-datetime` and an `ion-icon` and Ionic would take care of displaying/dismissing the datetime in a modal or popover. This would allow developers to not have to manage the modal state or add their own trigger buttons.

3. **There is no scroll indicator on the wheel picker.**

As noted in https://github.com/ionic-team/ionic-framework/issues/24477#issuecomment-1005972108, it sounds like if issues 1 and 2 are addressed then this issue is not as needed since developers could add a calendar picker as well as an input to type text into.

I am going to keep this issue open but will follow up when we have more to share. Thanks for everyone's patience!

**Comment by rdlabo** [comment]
New ion-datetime component is too cool!! But It spoils UI in certain situations.
For example: User will input birthday.

iOS's datePicker can use style '.wheels'.

```
datePicker.preferredDatePickerStyle = .wheels
```

I suggest ion-datetime added .wheels style. How do you think??

**Comment by andreas-aeschlimann** [comment]
> Hi everyone,
> 
> I wanted to give an update on the primary pain points noted here and the Ionic team's plan for addressing each:
> 
> 1. **When opening the time picker in the popover, it is not clear how to confirm the time because there are no buttons.**
> 
> As noted in [#24477 (comment)](https://github.com/ionic-team/ionic-framework/issues/24477#issuecomment-1005881804), we recommend splitting the date and time controls out for this use case. This follows a pattern commonly seen on iOS where the date and time pickers are split into separate form controls. Doing this allows developers to provide clear instructions on how to confirm date and time values.
> 
> 2. **Adding a button to open a datetime from an input requires a lot of code overhead to display the datetime in a modal.**
> 
> We are going to explore the idea of making this easier to manage via #24402. The idea here is that you could pass in `ion-datetime` and an `ion-icon` and Ionic would take care of displaying/dismissing the datetime in a modal or popover. This would allow developers to not have to manage the modal state or add their own trigger buttons.
> 
> 3. **There is no scroll indicator on the wheel picker.**
> 
> As noted in [#24477 (comment)](https://github.com/ionic-team/ionic-framework/issues/24477#issuecomment-1005972108), it sounds like if issues 1 and 2 are addressed then this issue is not as needed since developers could add a calendar picker as well as an input to type text into.
> 
> I am going to keep this issue open but will follow up when we have more to share. Thanks for everyone's patience!

Thanks @liamdebeasi.

So will there be an actual (classic) input option when a button is showed? That would solve the Desktop problem quite easily. The main problems on Desktop right now are the missing scroll indicator and buttons.

**Comment by liamdebeasi** [comment]
> New ion-datetime component is too cool!! But It spoils UI in certain situations. For example: User will input birthday.
> 
> iOS's datePicker can use style '.wheels'.
> 
> ```
> datePicker.preferredDatePickerStyle = .wheels
> ```
> 
> I suggest ion-datetime added .wheels style. How do you think??

@rdlabo This is something we would like to add. See https://github.com/ionic-team/ionic-framework/issues/23981.

-----

> Thanks @liamdebeasi.
> 
> So will there be an actual (classic) input option when a button is showed? That would solve the Desktop problem quite easily. The main problems on Desktop right now are the missing scroll indicator and buttons.


The exact API still needs to be discussed by the team, but the end result is that these changes will make it easier to present a datetime by clicking on a button inside of a text input.

**Comment by bkarv** [comment]
Just our 2cents feedback we have on ion-datetime from our users is the calendar day pick is very welcomed, thank you Ionic team for your hard efforts! 

However, when trying to choose the year or month the wheel is very awkward via a mouse especially if you do not have mouse-wheel. In some ways step backwards to the original wheel as the cursors selects the text and it just doesn't feel right on a desktop.

**1. Can we please have a quick fix to remove the cursor text-select on years and month?**

**2. Is there any reason why Ionic can't follow the similar year and month selections as you have done for days and like the other calendar widgets in Materials and ngx-Bootstrap?**


Cursor Selecting text in Desktop mode
<img width="346" alt="Screen Shot 2022-02-09 at 3 25 54 pm" src="https://user-images.githubusercontent.com/10624315/153121760-4fc687e4-a5b3-4fd7-99d5-226165fdce09.png">

Ideal Month and Year Pickers

<img width="298" alt="Screen Shot 2022-02-09 at 3 32 35 pm" src="https://user-images.githubusercontent.com/10624315/153122348-4fd98af6-da75-40ba-9b57-05d4e4db8fe6.png">

<img width="281" alt="Screen Shot 2022-02-09 at 3 30 20 pm" src="https://user-images.githubusercontent.com/10624315/153122138-5d953f4c-bbf2-4a6d-acdb-b3b4c6bf3679.png">

**Comment by liamdebeasi** [comment]
@bkarv Can you submit a separate issue for the text selection concern?

As for your second question, this was an intentional design decision to give datetime a consistent experience across iOS and Android.

**Comment by liamdebeasi** [comment]
Hi everyone,

I have some updates from the team. There are several concerns here, so I will address each one individually.

1. **The date picker looks odd in a form because it is so big.**

We did a couple things to make the default presentation look a bit better.
- We set a max-width of 350px on the calendar grid variant.
- We introduced the `size` property that lets developers make the datetime width be fixed or fill the entire width of the row it is in.

2. **There needs to be a way to easily show the date picker in a modal or popover.**

In Ionic 6.2 we introduced `ion-datetime-button` which lets you present a correctly sized/formatted datetime in a modal or popover.

Ionic 6.2 announcement blog: https://ionicframework.com/blog/announcing-ionic-v6-2/
Datetime Button documentation: https://ionicframework.com/docs/api/datetime-button

3. **When in a modal, there should be buttons to close the date picker.**

Confirmation buttons can be shown with the `showDefaultButtons` property.
`showDefaultButtons` documentation: https://ionicframework.com/docs/api/datetime#showing-confirmation-buttons

4. **Nested confirmation buttons should be shown when accessing the month/year picker in the calendar grid.**

Sean explained in https://github.com/ionic-team/ionic-framework/issues/24477#issuecomment-1005881804 that this is a UI pattern that we would like to avoid. See that comment for more information.

5. **There needs to be a way to use the wheel picker for selecting dates.**

In Ionic 6.2 we introduced the `preferWheel` property on `ion-datetime` which lets you select dates from a wheel picker. This feature also pairs well with `ion-datetime-button`.

`preferWheel` documentation: https://ionicframework.com/docs/api/datetime#wheel-style-pickers


--------

There are some other concerns that I believe I have addressed separately. As the main asks of the original post have been addressed, I am going to close this out. Please open a separate issue for any other concerns or requests you may have. Thank you!

**Comment by andreas-aeschlimann** [comment]
Thank you @liamdebeasi and the team for this great effort. I appreciate it a lot and I am sorry that my original post seemed so negative.

**Comment by liamdebeasi** [comment]
No worries at all, we did not take it that way. A date picker is a hard component to get right, so we appreciate all the feedback.
